### ISSUE TITLE
compact() doesn't reports if variable is defined after it

### ISSUE THREAD

**Issue by rentalhost** [kalessil/phpinspectionsea#112]
`compact()` should reports if variable doesn't was defined when it is executed, but it will not happen if you define the argumented variable after it. For instance:

```php
function test() {
    compact('user');
    $user = null;
}
```

`$user` was defined after `compact()`, but it is consided as valid by EA, when it should not be.

**Comment by kalessil** [comment]
I gave a try to fixing this, but end up with a very fragile solution. Though this case [has been documented](https://github.com/kalessil/phpinspectionsea/blob/master/docs/probable-bugs.md#compact-arguments-existence) instead.

**Comment by rentalhost** [comment]
Oh, it is bad. What was the major problem that you found? Maybe I can help to think in an alternative solution. I imagine that is something related to IDEA API.

**Comment by kalessil** [comment]
You are right, that's a limitation of the API: [here](https://github.com/kalessil/phpinspectionsea/blob/master/src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/apiUsage/CompactArgumentsInspector.java#L64) the scope being scanned for all used variables.

I simply can not say to IDE "Please only between scope entry point and the compact call".
I also limited with straightforward checking the lines associated with an expression.

And trying to simulate workflow is endless amount of work, no quick wins...

**Comment by kalessil** [comment]
I'd suggest just leave as it is - the inspection is doing great work though.

**Comment by rentalhost** [comment]
Yeah, it should solve a lot of current problems, but I still think that it is important because it should help when you are moving some code (this way was how I found this issue, I was reworking a code, moving something then I get a compact() error on PHP because EA doesn't warned me... yeah, I'm an EA dependent haha).

Well, even though for now you are not too worried about solving this problem - because I imagine there are several more important ones - maybe the case of keeping this topic open for future visits. A wontfix is an end of line for something that could be improved in the future.

So I'll leave a possible solution that I imagined: from your code, seems that you just get all variables declared on current scope. It's possible use the `findChildrenOfType()` to search by `compact()` and variables together? Then you could use de `Collection` index to do somehing like that:

```php
$var1 = true; // $var1 index 0
compact('var1', 'var2'); // compact() index 1
$var2 = true; // $var2 index 2
```

Then try to matches variables declared before current compact index like that (note: pseudo-code):

```java
for (Variable scopeObject : PsiTreeUtil.findChildrenOfType(scope, Variable.class | Compact.class)) {
	if (scopeObject instanceof Compact) {
	    final HashSet<String> variablesDeclaredUntilHere = new HashSet<>();

	    for (Variable scopeVariable : PsiTreeUtil.findChildrenOfType(scope, Variable.class)) {
	    	if (scopeVariableIndex >= scopeObjectIndex) {
	    		break;
	    	}

	        variablesDeclaredUntilHere.add(scopeVariable.getName());
	    }

	    // Here check if `variablesCompacted` was totally declared on `variablesDeclaredUntilHere`.
	}
}
```

**Comment by kalessil** [comment]
I'll re-check if the solution can be implemented.

**Comment by kalessil** [comment]
My congrats, the approach did work - the IDE has needed API ;)

**Comment by rentalhost** [comment]
Wow! It is almost magical! Thanks a lot! :)

**Comment by kalessil** [comment]
Indeed and you are welcome )
